Support named parameters in the `figure` shortcode

agriyakhetarpal opened the pull request:
## Description
This PR allows named parameters in the figure shortcode and is an attempt to unify the base Hugo shortcode at [https://github.com/gohugoio/hugo/blob/`@a1cb15e`/tpl/tplimpl/embedded/templates/_shortcodes/figure.html](https://github.com/gohugoio/hugo/blob/a1cb15e1cf9b7606d36552dc31d2ade613318c55/tpl/tplimpl/embedded/templates/_shortcodes/figure.html) with the theme's shortcode.

The change here allows the shortcode to be written in this manner: `{{< figure src="image.jpg" alt="description" >}}`, while preserving backwards compatibility as this is only in a code branch.

I've used [`IsNamedParams`](https://gohugo.io/methods/shortcode/isnamedparams/) to determine what form of shortcode we're using out of the two.

Based on https://gohugo.io/content-management/shortcodes/, the arguments inside the tag all use double-quoted strings, and I found that using single-quoted strings can cause errors that are a little difficult to debug based on Hugo's error messages, so I changed all instances to use double-quoted strings where applicable.

I added another example in the docs right above the shortcode on how to use this syntax. Also, the link to the image of the cute puppies was broken, so I added another one from Unsplash for our floofy friends :)

## Additional context

@goanpeca mentioned as a part of the work on https://github.com/scientific-python-translations/ that this would be nice to support, as Crowdin, as a part of the automations set up for the projects' websites, only understands the named-parameters syntax.

agriyakhetarpal commented:
Working well on https://deploy-preview-674--scientific-python-hugo-theme.netlify.app/shortcodes/#figure

stefanv reviewed (approved):
This change looks good to me.

We could also *only* support named parameters, we'd just need to fix up numpy.org, scipy.org, and all the SP websites. But, this is fine for now.

stefanv commented:
One last comment: I don't think we need the closing figure tags?

agriyakhetarpal commented:
> This change looks good to me.
> 
> We could also _only_ support named parameters, we'd just need to fix up numpy.org, scipy.org, and all the SP websites. But, this is fine for now.

Hi @stefanv, thank you for the review! Yes, I considered supporting only named parameters, but opted not to to preserve backwards compatibility.

> One last comment: I don't think we need the closing figure tags?

> Hi @stefanv, I think @agriyakhetarpal tried to work on that too but there were some issues.

Yes – I tried to make it work without closing figure tags, but I had an issue with the HTML character entities not being converted to their symbol counterparts (such as `&#34` for `"` and `/&gt` for `>`, and so on), as they were being rendered literally and breaking the shortcode, so I thought of leaving it so for now. I would be happy to explore this as a follow-up.

In the meantime, I've changed some more previously broken links for the cute puppies. I've confirmed that [the image I used](https://unsplash.com/photos/five-yellow-labrador-retriever-puppies-rfL-thiRzDs) is free to use under the Unsplash license. :)

stefanv commented:
> Yes – I tried to make it work without closing figure tags, but I had an issue with the HTML character entities not being converted to their symbol counterparts (such as `&#34` for `"` and `/&gt` for `>`, and so on), as they were being rendered literally and breaking the shortcode, so I thought of leaving it so for now. I would be happy to explore this as a follow-up.

If we are going to introduce a new syntax, it would be nice to have it "done and dusted" in one go. Do you think it would take long to explore whether we could use `{{< figure param=... /> }}`? 

> In the meantime, I've changed some more previously broken links for the cute puppies. I've confirmed that [the image I used](https://unsplash.com/photos/five-yellow-labrador-retriever-puppies-rfL-thiRzDs) is free to use under the Unsplash license. :)

I happened to check that too earlier today ;)

agriyakhetarpal commented:
> If we are going to introduce a new syntax, it would be nice to have it "done and dusted" in one go. Do you think it would take long to explore whether we could use `{{< figure param=... /> }}`?

Yes, this is what's broken. I'll take a look, though!

stefanv commented:
I wonder if the closing tag is required because we access `.Inner`? Or is `.Inner` only available when there is a closing tag? (That's be a bit odd.)

agriyakhetarpal commented:
Yes, that's indeed the case (@goanpeca and I were discussing yesterday). I'm trying to see if we can make it work without it!

agriyakhetarpal commented:
Okay, yeah, I don't think this is going to work, based on what I understand from https://discourse.gohugo.io/t/what-is-meant-by-closed-or-self-closed/44785 and https://gohugo.io/templates/shortcode/#inner. `.Inner` is available only with a closing tag.

I think we could:
- create a `figure-single.html` to handle the self-closing syntax specifically that would simply collect parameters and pass them to the regular figure shortcode as a partial, but that would be a new shortcode and I think it would confuse users.
	- maybe, we could rename the current `figure.html` to `figure-base.html` (and not document the `figure-base` shortcode), go ahead with this solution, and name that `figure.html`? It could work as nothing would be exposed to the user, but I think it confuses maintainers down the line.
- The other way is to use Hugo's own figure shortcode and adapt it to include the extra parameters we have as the self-closing syntax works there – but that breaks compatibility.

I would probably opt for none of these at the moment, at least for this PR :)

stefanv commented:
Having two syntaxes is confusing. Let's just move over to the newly proposed tag, make sure the change is mentioned in the release notes, and port the other websites that depend on it (there aren't that many).

stefanv commented:
(If, that is, I understood correctly that removing `.Inner` will fix the problem.)

agriyakhetarpal commented:
> Having two syntaxes is confusing. Let's just move over to the newly proposed tag, make sure the change is mentioned in the release notes, and port the other websites that depend on it (there aren't that many).

> (If, that is, I understood correctly that removing .Inner will fix the problem.)

Yes, I get this, but I think it would also be odd to have just the `{{< figure >}}` shortcode with the new named parameters syntax + dropping `.Inner` to support self-closing tags for it, and not any of the other shortcodes that are also using `.Inner`? Should I aim to port the rest of them, too? If we do, then it breaks a lot more content downstream (which I have no issues fixing, though!).

stefanv commented:
I suspect you're going to run into issues with some of the other shortcodes, such as cards.

But, I think you are right that we should be consistent.

Can we use the same pattern as here to make inline/in-body usage possible for those?

agriyakhetarpal commented:
Yes, I think it should be possible. I think we can also keep it as is for shortcodes that are tricky to port or where it doesn't make sense to change. For example, the `admonition` shortcode:

```markdown
{{< admonition hint >}}
This is an hint admonition.
{{< /admonition >}}
```

seems perfectly fine to me, as `{{< admonition hint  "This is an hint admonition." />}}` looks a little off, and would not work for complex uses where someone may want a mixture of multiline strings + images + tables, and whatnot within the admonition... This is an example where the current one works well.

Here's a list of the shortcodes that I think can be ported:

1. `button`:

It would go from

```markdown
{{< button success >}}
label: Success
link: http://example.com/
{{< /button >}}
```

to

```markdown
{{< button success label = "Success" link = "http://example.com/" />}}
```

Similarly,

2. Dropdown
3. Image

<hr>

The ones I think are fine as is:

1. Admonition
2. Card
3. Dropdown (as it's possible to use multiline strings + images + complex Markdown)
5. Details
6. Field List (uses TOML tables)
7. Grid (same as Field List)
8. Sidebar (as it supports nested shortcodes based on the provided example)
9. Tabs

<hr>

The ones below don't use `.Inner` at all currently and are thus fine as is:

1. include-code
2. include-html
3. include-md
5. include-raw
6. include
7. toctree

<hr>

I can port the `button`, `dropdown`, and the `image` shortcodes, unless you suggest otherwise! Another option is to just port `figure` and `image` for now, as both of them are closely related in terms of functionality.

stefanv commented:
Yes, I think the rule would be: anything that calls for *parameters* in the body, we can port. With the exception of cards, because those parameters are actually content :roll_eyes: :grimacing:

agriyakhetarpal commented:
I'll finish up later today, thanks!

goanpeca commented:
>Having two syntaxes is confusing. Let's just move over to the newly proposed tag, make sure the change is mentioned in the release notes

Indeed it might be better to have just one. Thanks for the commentary and suggestions @stefanv!

I wanted to ask if we could limit the scope of the PR to what @agriyakhetarpal initially worked on and in a subsequent PR make all the additional changes.

We (quansight) are finishing the work for the scientific python translations grant and it would help of we can deploy numpy l.org soon which is currently blocked due to some issues with crowdin (the translations platform) not understanding the current shortcode syntax.

I have already migrated the numpy figures over at https://github.com/numpy/numpy.org/pull/859

I can help migrating other sites as well.

stefanv commented:
Fine to port one at a time, but let's port to the desired format directly.

stefanv commented:
(Also, a bit nervous about partial porting of shortcodes, and ending in an inconsistent state.)

goanpeca commented:
@agriyakhetarpal is this ready for review now?

---

>(Also, a bit nervous about partial porting of shortcodes, and ending in an inconsistent state.)

@stefanv I understand, but I can work on the port of the rest as soon as we can redeploy the numpy website.

agriyakhetarpal commented:
Hi @goanpeca, I didn't get time to work on this PR further as much as I would have loved to, unfortunately.

In the interest of time and for the completion your and Melissa's work on the translations front, I think we could explore if I can hand it off from here to you? If you wish to build on this, I shall be happy to share access to my fork with you, so that you can push directly to this PR and complete the rest.

The remaining work here is:
- use named parameters only (drop `IsNamedParams` and conditions that rely on it)
- support the image, button, and dropdown shortcodes similarly

Please let me know if this works!

agriyakhetarpal commented:
Also, considering @stefanv's concern around partial porting of shortcodes, I'd assume that we wouldn't have a new release of the theme until all relevant ones are done – I would also be curious to know what the policy for the numpy.org and scipy.org maintainers is, i.e., whether they would be okay to rely on an unreleased change as a part of their theme submodule update.

goanpeca commented:
Hi @agriyakhetarpal 

>If you wish to build on this, I shall be happy to share access to my fork with you, so that you can push directly to this PR and complete the rest.

Yes that works, thanks a lot :) 

---

>I would also be curious to know what the policy for the numpy.org and scipy.org maintainers is, i.e., whether they would be okay to rely on an unreleased change as a part of their theme submodule update for their deployments.

I think it would be fine to update the submodule to point to the specific commit that fixes the issue, and that would be fine.

agriyakhetarpal commented:
Awesome, thanks!

I've extended an invite for my fork to you. Please LMK if it doesn't work or if it hasn't shown up!

goanpeca commented:
Accepted! Thanks

goanpeca commented:
Hi @stefanv, I will continue working on this migration. Just to double check, we want to only support the new syntax with no closing tags for the shortcodes:

- figure
- button
- image
- dropdown

?

goanpeca commented:
Hello team, thanks @agriyakhetarpal for the initial work.

I updated the shortcodes for

- figure
- button
- image
- dropdown

To make them work inline only with bno closing tag.

goanpeca commented:
![Screenshot 2025-05-12 at 21-07-49 Scientific Python Hugo Theme documentation - Shortcodes](https://github.com/user-attachments/assets/bb988592-29a0-42d9-8996-7a0fef9a3e29)

agriyakhetarpal commented:
Okay, the site is building and working well locally 🎉 but as the tests in CI run the SciPy website, they will show up as failed because it hasn't been updated for the new syntax yet.

@goanpeca and I have discussed this today, and I've sent out a list of all the websites that need to be ported with the new syntax once this PR gets merged and makes it to a release, and when the new release is used for those websites. The change for the figure shortcode in specific was tested out in numpy/numpy.org#859, and it worked well. Many thanks for finishing the rest of the shortcodes!

The list is as follows (please let me know if I missed any!)

#### Scientific Python core websites
- [ ] [learn.scientific-python.org](http://learn.scientific-python.org/)
- [ ] [scientific-python.org](http://scientific-python.org/)
- [ ] [tools.scientific-python.org](http://tools.scientific-python.org/)
- [ ] [blog.scientific-python.org](http://blog.scientific-python.org/)
- [ ] [devstats.scientific-python.org](http://devstats.scientific-python.org/)
- [ ] [specs.scientific-python.org](http://specs.scientific-python.org/)

#### Downstream users of the theme
- [ ] [numpy.org](http://numpy.org/)
- [ ] [scipy.org](http://scipy.org/)
- [ ] [pybamm.org](http://pybamm.org/)
- [ ] https://pysathq.github.io/

<hr>

Also, post the porting of the shortcodes I managed to figure out the problem where the closing tags (`>}}`) for the shortcodes were being displayed as blockquotes – I realised that this happened only on the shortcodes page, and the single source of truth for what's being modified was in the `render_shortcode_docs.py` script. I've put out a fix in cd16575aa4b1aab7e91b1f2d2ee6c9b6b385c383.

Overall, we now have the "figure", "image", "dropdown", and "button" shortcodes supporting named parameters, so this PR is ready for review/merging. Thank you!

agriyakhetarpal commented:
Hi @goanpeca, I think we should forego the changes to the "dropdown" shortcode; I think we made an oversight. In another PR that I was working on just now (#675), I noticed that dropdowns can have content in the body parameter, which doesn't make sense to inline as inline parameters are not processed – which means that if one were to add something like a code block or a complex multiline string, it wouldn't work well.

Originally, in https://github.com/scientific-python/scientific-python-hugo-theme/pull/674#issuecomment-2840813416, I made the mistake where I added the "dropdown" shortcode to both headings – in the "ones to modify" and "those to keep as-is", when we should have kept it only in the as-is heading 😅 I've updated that comment, as I think it was the cause of some confusion! Similar to Stefan's comment in https://github.com/scientific-python/scientific-python-hugo-theme/pull/674#issuecomment-2840819005 about cards, one of the parameters to the dropdown(s) is actually content, too.

I converted the PR to a draft again so that I can undo these changes specifically, so that we support inline parameters for only the `image`, `figure`, and `button` shortcodes, and I will update the PR description, etc., again.

goanpeca commented:
Hello team, I created PRs with the updated syntax in preparation for these changes.

- scipy: https://github.com/scipy/scipy.org/pull/648
- numpy: https://github.com/numpy/numpy.org/pull/859
- pybann: https://github.com/pybamm-team/pybamm.org/pull/283
- learn.scientific-python.org: no changes needed.
- scientific-python.org: https://github.com/scientific-python/scientific-python.org/pull/768
- tools.scientific-python.org: https://github.com/scientific-python/tools.scientific-python.org/pull/72
- blog.scientific-python.org: no changes needed.
- devstats.scientific-python.org: no changes needed.
- specs.scientific-python.org: https://github.com/scientific-python/specs/pull/394

stefanv reviewed (commented):
Looks good to me, overall.

It may be worth adding a note for developers on when to use keywords and when to use TOML in the body.

agriyakhetarpal commented on layouts/shortcodes/figure.html:
```diff
@@ -55,7 +52,7 @@
 {{- end }}
 <p><span class="caption-text">
-{{- $figure.caption | markdownify -}}
+{{- $figure.caption | markdownify | plainify -}}
```
Okay, after careful consideration, I am deciding _not_ to support shortcodes of the following form:

```markdown
{{< figure
    src="https://images.unsplash.com/photo-1546238232-20216dec9f72"
    alt="Cute puppies"
    height=200
    width=400
    attribution="Figure Credits: Cute puppies image from unsplash.com"
    attributionlink="https://images.unsplash.com/photo-1546238232-20216dec9f72"
    caption=`
            A figure with a caption split across multiple lines.
            <br>
            <br>
            You can use **Markdown** _formatting_ within the caption text and use it to
            fit the needs of your content.
            <br>
            As always, be sure to credit the image source!
            `
>}}
```

If someone wants to add a caption with multiple lines, instead of using backticks, they can simply use `<br>` tags, and split them as they please. This is the same as what is being done in the `legend` parameter in the example only a few lines above! We discussed in our chat already that this is a niche use case, so the above shortcode for them should look like this:

```markdown
{{< figure
    src="https://images.unsplash.com/photo-1546238232-20216dec9f72"
    alt="Cute puppies"
    height=200
    width=400
    attribution="Figure Credits: Cute puppies image from unsplash.com"
    attributionlink="https://images.unsplash.com/photo-1546238232-20216dec9f72"
    caption="A figure with a caption split across multiple lines.<br><br>You can use **Markdown** _formatting_ within the caption text and use it to fit the needs of your content.<br>As always, be sure to credit the image source!"
>}}
```

which works right now. I've dropped the `plainify` as there's no longer any need for it, and documented this example in faca829d8d70fa56ca479f2af820cff2cedd9383.

agriyakhetarpal commented:
> Looks good to me, overall.
> 
> It may be worth adding a note for developers on when to use keywords and when to use TOML in the body.

Thanks for the review! I'm afraid I don't understand this, though – could you please clarify? Do you mean we should highlight the shortcodes that will use the new syntax, and the ones that will stick to the old one?

agriyakhetarpal commented:
Yes, the builds will fail (xref: https://github.com/scientific-python/scientific-python-hugo-theme/pull/674#issuecomment-2874791086, https://github.com/scientific-python/scientific-python-hugo-theme/pull/674#issuecomment-2877386646) as the tests build the SciPy website in CI. They will pass here when https://github.com/scipy/scipy.org/pull/648 is merged (it requires the theme submodule bump).

Here is a list of @goanpeca's PRs for all downstream websites: https://github.com/scientific-python/scientific-python-hugo-theme/pull/674#issuecomment-2928357436

P.S. I wonder if we could make our downstream testing easier with Pixi tasks. It's surely worth an experiment...

stefanv commented:
Yes, we should try the pixi experiment!

Re: developer guidelines also see #616

Thanks, I think we're good to go here then, and to tag a new release! 🎉

agriyakhetarpal commented:
@goanpeca, we'll release v0.22 in a while (finally!). When you get a chance, could you please rebase the downstream PRs for all the website and bump the theme version in them? Thank you!
